# Purpose of this notebook
- Use Dense to solve cifar10

# Import

In [1]:
from tensorflow import keras # type: ignore
from tensorflow.keras import layers # type: ignore
from tensorflow.keras.datasets import cifar10 # type: ignore

# Load data

In [2]:
(x_train, y_train), (x_test, y_test) = cifar10.load_data()
print(x_train.shape)
print(y_train.shape)
print(x_test.shape)
print(y_test.shape)

170498071/170498071 ━━━━━━━━━━━━━━━━━━━━ 38s 0us/step
(50000, 32, 32, 3)
(50000, 1)
(10000, 32, 32, 3)
(10000, 1)


# Before normalize, range in [0, 255]

In [7]:
print(x_train[0][0][0])
print(x_test[0][0][0])

[59 62 63]
[158 112  49]


# Normalize
- Note: Run following cell only once

In [8]:
x_train = x_train / 255
x_test = x_test / 255

# After normalize, range in [0, 1]

In [9]:
print(x_train[0][0][0])
print(x_test[0][0][0])

[0.23137255 0.24313725 0.24705882]
[0.61960784 0.43921569 0.19215686]


# Keras Sequential API

In [11]:
model = keras.Sequential([
    layers.Flatten(input_shape=(32, 32, 3), name='flatten_layer'),
    layers.Dense(512, activation='relu', name='first_layer'),
    layers.Dense(256, activation='relu', name='second_layer'),
    layers.Dense(10, name='output_layer'),
], name="sequential_model")

# model.summary 

In [12]:
print(model.summary())

Model: "sequential_model"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten_layer (Flatten)         │ (None, 3072)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ first_layer (Dense)             │ (None, 512)            │     1,573,376 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ second_layer (Dense)            │ (None, 256)            │       131,328 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ output_layer (Dense)            │ (None, 10)             │         2,570 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 1,707,274 (6.51 MB)

 Trainable params: 1,707,274 (6.51 MB)

 Non-trainable params: 0 (0.00 B)

None


# model.compile

In [13]:
model.compile(
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    optimizer=keras.optimizers.Adam(learning_rate=0.001),
    metrics=["accuracy"],
)

# Note
- If output layer `layers.Dense(10)` then use `from_logits=True`
- If output layer `layers.Dense(10, activation='softmax')` then use `from_logits=False`, or just simply remove `from_logits=False`

# Train the model using model.fit

In [14]:
model.fit(x_train, y_train, batch_size=32, epochs=5, verbose=2)

Epoch 1/5
1563/1563 - 22s - 14ms/step - accuracy: 0.3298 - loss: 1.8567
Epoch 2/5
1563/1563 - 21s - 13ms/step - accuracy: 0.4067 - loss: 1.6603
Epoch 3/5
1563/1563 - 21s - 13ms/step - accuracy: 0.4359 - loss: 1.5838
Epoch 4/5
1563/1563 - 22s - 14ms/step - accuracy: 0.4511 - loss: 1.5347
Epoch 5/5
1563/1563 - 21s - 13ms/step - accuracy: 0.4650 - loss: 1.4985


# Note
- Wait model.fit above run in 1m46s (long time)
- But just achieve accuracy 0.4650 (less than 50%)

# Conclusion
- Use Dense to solve image classification is inefficient
- => Use CNN instead (see 10.ipynb)

# Evaluate test data using model.evaluate

In [15]:
model.evaluate(x_test, y_test, batch_size=32, verbose=2)

313/313 - 0s - 1ms/step - accuracy: 0.4591 - loss: 1.5021


[1.5021075010299683, 0.45910000801086426]